packages/react-components/react-combobox/src/hooks/useComboboxFilter.tsx

```diff
@@ -0,0 +1,83 @@
+import * as React from 'react';
+
+import type { ComboboxProps } from '../Combobox';
+import { Option } from '../Option';
+
+type UseComboboxFilterConfig<T extends { children: React.ReactNode; value: string } | string> = {
+  /** Provides a custom filter for the option. */
+  filter?: (optionText: string, query: string) => boolean;
+
+  /** Provides a custom message to display when there are no options. */
+  noOptionsMessage?: React.ReactNode;
+
+  /** Provides a way to map an option object to a React key. By default, "value" is used. */
+  optionToReactKey?: (option: T) => string;
+
+  /** Provides a way to map an option object to a text used for search. By default, "value" is used. */
+  optionToText?: (option: T) => string;
+
+  /** Provides a custom render for the option. */
+  renderOption?: (option: T) => JSX.Element;
+};
+
+function defaultFilter(optionText: string, query: string) {
+  if (query === '') {
+    return true;
+  }
+
+  return optionText.toLowerCase().includes(query.toLowerCase());
+}
+
+function defaultToString(option: string | { value: string }) {
+  return typeof option === 'string' ? option : option.value;
+}
+
+export function useComboboxFilter<T extends { children: React.ReactNode; value: string } | string>(
+  options: T[],
+  config: UseComboboxFilterConfig<T>,
+) {
+  const {
+    filter = defaultFilter,
+    noOptionsMessage = "We couldn't find any matches.",
+    optionToReactKey = defaultToString,
+    optionToText = defaultToString,
+
+    renderOption = (option: T) => {
```

Comment:
I think each option should have its own renderer, since asymmetric options are not uncommon in apps

Reply:
As discussed in offline, currently we have some flexibility there and it depends on what customers would like to achieve.

For example, we have following data structure:

```ts
const data = [
  { type: 'fruit', value: 'Apple' },
  { type: 'fruit', value: 'Banana' },
  { type: 'vegetable', value: 'Carrot' },
];
```

Now, we can already do:

```ts
useComboboxFilter({
  renderOption: option => {
    if (option.type === 'fruit') {
      /* ... */
    }
  },
});
```

If we will add a possibility to have render functions for each option:
- there will be two ways of doing things
- an additional iteration on data is required as functions usually don't come from data layer

```ts
const options = data.map(option => {
  if (option.type === 'fruit') {
    return { ...option, render: renderFruit };
  }

  /* ... */
});
```

Based on the previous experience with Teams and `Dropdown` there, users preferred `renderItem()` approach. Anyway, we can reconsider later based on feedback.